File: app/src/main/java/com/webianks/hatkemessenger/activities/SmsDetailedView.java
package com.webianks.hatkemessenger.activities;

import android.Manifest;
import android.app.PendingIntent;
import android.content.BroadcastReceiver;
import android.content.Intent;
import android.content.IntentFilter;
import android.content.pm.PackageManager;
import android.database.Cursor;
import android.os.Bundle;
import android.support.v4.app.ActivityCompat;
import android.support.v4.app.LoaderManager;
import android.support.v4.content.ContextCompat;
import android.support.v4.content.CursorLoader;
import android.support.v4.content.Loader;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.telephony.SmsManager;
import android.view.MenuItem;
import android.view.View;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.Toast;

import com.webianks.hatkemessenger.R;
import com.webianks.hatkemessenger.adapters.SingleGroupAdapter;
import com.webianks.hatkemessenger.constants.Constants;
import com.webianks.hatkemessenger.constants.SmsContract;
import com.webianks.hatkemessenger.receivers.DeliverReceiver;
import com.webianks.hatkemessenger.receivers.SentReceiver;
import com.webianks.hatkemessenger.services.UpdateSMSService;

public class SmsDetailedView extends AppCompatActivity implements
        LoaderManager.LoaderCallbacks<Cursor>, View.OnClickListener {

    private String contact;
    private SingleGroupAdapter singleGroupAdapter;
    private RecyclerView recyclerView;
    private EditText etMessage;
    private ImageView btSend;
    private String message;
    private boolean from_reciever;
    private long _Id;
    private int color;
    private String read = "1";


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_sms_detailed_view);
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);

        init();
    }


    private void init() {

        Intent intent = getIntent();


        contact = intent.getStringExtra(Constants.CONTACT_NAME);
        _Id = intent.getLongExtra(Constants.SMS_ID,-123);
        color = intent.getIntExtra(Constants.COLOR,0);
        read = intent.getStringExtra(Constants.READ);

        from_reciever = intent.getBooleanExtra(Constants.FROM_SMS_RECIEVER, false);

        if (getSupportActionBar() != null)
            getSupportActionBar().setTitle(contact);

        recyclerView = (RecyclerView) findViewById(R.id.recyclerview);
        LinearLayoutManager linearLayoutManager = new LinearLayoutManager(this);
        linearLayoutManager.setReverseLayout(true);
        recyclerView.setLayoutManager(linearLayoutManager);

        etMessage = (EditText) findViewById(R.id.etMessage);
        btSend = (ImageView) findViewById(R.id.btSend);

        btSend.setOnClickListener(this);

        setRecyclerView(null);

        if (read!=null && read.equals("0"))
            setReadSMS();
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {

        switch (item.getItemId()) {
            case android.R.id.home:

                if (from_reciever)
                    startActivity(new Intent(this, MainActivity.class));

                finish();
                break;
        }

        return super.onOptionsItemSelected(item);
    }


    private void setRecyclerView(Cursor cursor) {
        singleGroupAdapter = new SingleGroupAdapter(this, cursor,color);
        recyclerView.setAdapter(singleGroupAdapter);
    }

    @Override
    protected void onResume() {
        super.onResume();
        getSupportLoaderManager().initLoader(Constants.CONVERSATION_LOADER, null, this);
    }

    @Override
    public Loader<Cursor> onCreateLoader(int id, Bundle args) {

        String[] selectionArgs = {contact};

        return new CursorLoader(this,
                SmsContract.ALL_SMS_URI,
                null,
                SmsContract.SMS_SELECTION,
                selectionArgs,
                null);
    }

    @Override
    public void onLoadFinished(Loader<Cursor> loader, Cursor cursor) {

        if (cursor != null && cursor.getCount() > 0) {
            singleGroupAdapter.swapCursor(cursor);
        } else {
            //no sms
        }
    }


    private void setReadSMS() {

        Intent intent = new Intent(this, UpdateSMSService.class);
        intent.putExtra("id", _Id);
        startService(intent);
    }

    @Override
    public void onLoaderReset(Loader<Cursor> loader) {
        singleGroupAdapter.swapCursor(null);
    }

    @Override
    public void onClick(View view) {

        switch (view.getId()) {
            case R.id.btSend:
                sendSMSMessage();
                break;
        }
    }

    protected void sendSMSMessage() {

        message = etMessage.getText().toString();

        if (message!=null && message.trim().length()>0)
            requestPermisions();
        else
            etMessage.setError(getString(R.string.please_write_message));

    }

    private void requestPermisions() {

        if (ContextCompat.checkSelfPermission(this,
                Manifest.permission.SEND_SMS)
                != PackageManager.PERMISSION_GRANTED) {
            if (ActivityCompat.shouldShowRequestPermissionRationale(this,
                    Manifest.permission.SEND_SMS)) {
            } else {
                ActivityCompat.requestPermissions(this,
                        new String[]{Manifest.permission.SEND_SMS},
                        Constants.MY_PERMISSIONS_REQUEST_SEND_SMS);
            }
        }else{
            sendSMSNow();
        }

    }

    @Override
    public void onRequestPermissionsResult(int requestCode, String permissions[], int[] grantResults) {
        switch (requestCode) {
            case Constants.MY_PERMISSIONS_REQUEST_SEND_SMS: {
                if (grantResults.length > 0
                        && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                    sendSMSNow();
                } else {
                    Toast.makeText(getApplicationContext(),
                            "SMS failed, please try again.", Toast.LENGTH_LONG).show();
                    return;
                }
            }
        }
    }

    private void sendSMSNow() {

        BroadcastReceiver sendBroadcastReceiver = new SentReceiver();
        BroadcastReceiver deliveryBroadcastReciever = new DeliverReceiver();

        String SENT = "SMS_SENT";
        String DELIVERED = "SMS_DELIVERED";

        PendingIntent sentPI = PendingIntent.getBroadcast(this, 0, new Intent(SENT), 0);
        PendingIntent deliveredPI = PendingIntent.getBroadcast(this, 0, new Intent(DELIVERED), 0);

        registerReceiver(sendBroadcastReceiver, new IntentFilter(SENT));
        registerReceiver(deliveryBroadcastReciever, new IntentFilter(DELIVERED));

        try {
            SmsManager sms = SmsManager.getDefault();
            sms.sendTextMessage(contact, null, message, sentPI, deliveredPI);
        }catch (Exception e){
            Toast.makeText(this,getString(R.string.cant_send),Toast.LENGTH_SHORT).show();

        }
    }

    @Override
    public void onBackPressed() {

        if (from_reciever) {
            startActivity(new Intent(this, MainActivity.class));
        } else
            super.onBackPressed();
    }
}

File: app/src/main/java/com/webianks/hatkemessenger/activities/SmsDetailedView.kt
package com.webianks.hatkemessenger.activities

import android.Manifest
import android.app.PendingIntent
import android.content.BroadcastReceiver
import android.content.Intent
import android.content.IntentFilter
import android.content.pm.PackageManager
import android.database.Cursor
import android.os.Bundle
import android.telephony.SmsManager
import android.view.MenuItem
import android.view.View
import android.widget.EditText
import android.widget.ImageView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import androidx.loader.app.LoaderManager
import androidx.loader.content.CursorLoader
import androidx.loader.content.Loader
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.webianks.hatkemessenger.R
import com.webianks.hatkemessenger.adapters.SingleGroupAdapter
import com.webianks.hatkemessenger.constants.Constants
import com.webianks.hatkemessenger.constants.SmsContract
import com.webianks.hatkemessenger.receivers.DeliverReceiver
import com.webianks.hatkemessenger.receivers.SentReceiver
import com.webianks.hatkemessenger.services.UpdateSMSService

class SmsDetailedView : AppCompatActivity(),
        LoaderManager.LoaderCallbacks<Cursor?>,
        View.OnClickListener {

    private var contact: String? = null
    private var savedContactName: String? = null
    private var singleGroupAdapter: SingleGroupAdapter? = null
    private var recyclerView: RecyclerView? = null
    private var etMessage: EditText? = null
    private var btSend: ImageView? = null
    private var message: String? = null
    private var from_reciever = false

    private var _Id: Long = 0
    private var color = 0
    private var read: String? = "1"

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_sms_detailed_view)
        supportActionBar!!.setDisplayHomeAsUpEnabled(true)
        init()
    }

    private fun init() {
        val intent = intent
        contact = intent.getStringExtra(Constants.CONTACT_NAME)
        savedContactName = intent.getStringExtra(Constants.SAVED_CONTACT_NAME)
        _Id = intent.getLongExtra(Constants.SMS_ID, -123)
        color = intent.getIntExtra(Constants.COLOR, 0)
        read = intent.getStringExtra(Constants.READ)
        from_reciever = intent.getBooleanExtra(Constants.FROM_SMS_RECIEVER, false)
        if (supportActionBar != null) {
            if (savedContactName == null)
                supportActionBar!!.setTitle(contact)
            else
                supportActionBar!!.setTitle(savedContactName)
        }
        recyclerView = findViewById(R.id.recyclerview)
        val linearLayoutManager = LinearLayoutManager(this)
        linearLayoutManager.reverseLayout = true
        recyclerView?.layoutManager = linearLayoutManager
        etMessage = findViewById(R.id.etMessage)
        btSend = findViewById(R.id.btSend)
        btSend?.setOnClickListener(this)
        setRecyclerView(null)
        if (read != null && read == "0") setReadSMS()
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == android.R.id.home) {
            if (from_reciever) startActivity(Intent(this, MainActivity::class.java))
            finish()
        }
        return super.onOptionsItemSelected(item)
    }


    private fun setRecyclerView(cursor: Cursor?) {
        singleGroupAdapter = SingleGroupAdapter(this, cursor, color, savedContactName)
        recyclerView!!.adapter = singleGroupAdapter
    }

    override fun onResume() {
        super.onResume()
        LoaderManager.getInstance(this).initLoader(Constants.CONVERSATION_LOADER, null, this)
    }

    override fun onCreateLoader(id: Int, args: Bundle?): Loader<Cursor?> {
        val selectionArgs = arrayOf(contact)
        return CursorLoader(this,
                SmsContract.ALL_SMS_URI,
                null,
                SmsContract.SMS_SELECTION,
                selectionArgs,
                null)
    }

    override fun onLoadFinished(loader: Loader<Cursor?>, cursor: Cursor?) {
        if (cursor != null && cursor.count > 0) {
            singleGroupAdapter!!.swapCursor(cursor)
        } //no sms
    }

    private fun setReadSMS() {
        val intent = Intent(this, UpdateSMSService::class.java)
        intent.putExtra("id", _Id)
        startService(intent)
    }

    override fun onLoaderReset(loader: Loader<Cursor?>) {
        singleGroupAdapter!!.swapCursor(null)
    }

    override fun onClick(view: View) {
        if (view.id == R.id.btSend) {
            sendSMSMessage()
        }
    }

    protected fun sendSMSMessage() {
        message = etMessage!!.text.toString()
        if (message!!.trim { it <= ' ' }.isNotEmpty())
            requestPermissions() else etMessage!!.error = getString(R.string.please_write_message)
    }

    private fun requestPermissions() = if (ContextCompat.checkSelfPermission(this,
                    Manifest.permission.SEND_SMS)
            != PackageManager.PERMISSION_GRANTED) {
        if (ActivityCompat.shouldShowRequestPermissionRationale(this,
                        Manifest.permission.SEND_SMS)) {
        } else {
            ActivityCompat.requestPermissions(this, arrayOf(Manifest.permission.SEND_SMS),
                    Constants.MY_PERMISSIONS_REQUEST_SEND_SMS)
        }
    } else {
        sendSMSNow()
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        if (requestCode == Constants.MY_PERMISSIONS_REQUEST_SEND_SMS) {
            if (grantResults.isNotEmpty()
                    && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                sendSMSNow()
            } else {
                Toast.makeText(applicationContext,
                        "SMS failed, please try again.", Toast.LENGTH_LONG).show()
            }
        }
    }

    private fun sendSMSNow() {
        val sendBroadcastReceiver: BroadcastReceiver = SentReceiver()
        val deliveryBroadcastReceiver: BroadcastReceiver = DeliverReceiver()
        val SENT = "SMS_SENT"
        val DELIVERED = "SMS_DELIVERED"
        val sentPI = PendingIntent.getBroadcast(this, 0, Intent(SENT), 0)
        val deliveredPI = PendingIntent.getBroadcast(this, 0, Intent(DELIVERED), 0)
        registerReceiver(sendBroadcastReceiver, IntentFilter(SENT))
        registerReceiver(deliveryBroadcastReceiver, IntentFilter(DELIVERED))
        try {
            val sms = SmsManager.getDefault()
            sms.sendTextMessage(contact, null, message, sentPI, deliveredPI)
        } catch (e: Exception) {
            Toast.makeText(this, getString(R.string.cant_send), Toast.LENGTH_SHORT).show()
        }
    }

    override fun onBackPressed() {
        if (from_reciever) {
            startActivity(Intent(this, MainActivity::class.java))
        } else super.onBackPressed()
    }
}